src/compiler/scala/tools/nsc/typechecker/Infer.scala

```diff
@@ -282,8 +282,10 @@ trait Infer extends Checkable {
             catch { case ex: MalformedType => malformed(ex, pre memberType underlyingSymbol(sym)) }
           )
           tree setSymbol sym1 setType (
-            pre match {
-              case _: SuperType => owntype map (tp => if (tp eq pre) site.symbol.thisType else tp)
+          pre match {
+              case _: SuperType =>
```

Comment:
Since the `then` case is also the default, perhaps we can fold the `if` condition into the pattern: 

```Scala
case _: SuperType if (sym.isConstructor || owntype.isInstanceOf[OverloadedType]) =>
```
This could better show that it is restricting the use of the "expensive" case.